File: android/app/src/main/java/com/lilru/liftr/ui/profile/PrFormatting.kt
package com.lilru.liftr.ui.profile

import com.lilru.liftr.ui.add.AddCardioActivity
import com.lilru.liftr.ui.add.AddSportType
import java.util.Locale
import kotlin.math.max
import kotlin.math.roundToInt

object PrFormatting {
    fun activityLabel(kind: String, label: String): String {
        val k = kind.lowercase(Locale.US)
        val wire = label.lowercase(Locale.US)
        if (k == "cardio") {
            AddCardioActivity.entries.find { it.wire == wire }?.let { return cardioTypeLabel(it) }
            if (wire == "cardio") return "Cardio"
            return snakeToTitle(wire)
        }
        if (k == "sport") {
            AddSportType.entries.find { it.wire == wire }?.let {
                return sportTypeLabel(it)
            }
            return snakeToTitle(wire)
        }
        return label
    }

    fun activityIcon(kind: String, label: String): String {
        val k = kind.lowercase(Locale.US)
        val wire = label.lowercase(Locale.US)
        return when (k) {
            "strength" -> "💪"
            "cardio" -> cardioIcon(wire)
            "sport" -> sportIcon(wire)
            else -> ""
        }
    }

    fun prettyMetricName(metric: String, kind: String = "", label: String = ""): String {
        val m = metric.lowercase(Locale.US)
        return when (m) {
            "max_hr", "max_avg_hr" -> "Max HR"
            "longest_duration_sec", "max_moving_time_sec" -> "Longest duration"
            "longest_distance_km", "max_total_distance_km" -> "Longest distance"
            "fastest_pace_sec_per_km" -> "Fastest pace"
            "max_elevation_m", "max_vertical_drop_m", "max_total_vertical_m" -> "Max elevation"
            "est_1rm_kg" -> "Estimated 1RM"
            "max_weight_kg" -> "Max weight"
            "best_set_volume_kg" -> "Best set volume"
            "max_reps" -> "Max reps"
            "max_cadence_rpm" -> "Max cadence"
            "max_watts_avg" -> "Max watts"
            "min_split_sec_per_500m" -> "Best split"
            "max_goals" -> "Max goals"
            "max_assists" -> "Max assists"
            "max_shots_on_target" -> "Max shots on target"
            "max_shots_on_goal" -> "Max shots on goal"
            "max_saves" -> "Max saves"
            "max_tackles", "max_tackles_made" -> "Max tackles"
            "max_yellow" -> "Max yellow cards"
            "max_red" -> "Max red cards"
            "max_points" -> "Max points"
            "max_rebounds" -> "Max rebounds"
            "max_steals" -> "Max steals"
            "max_blocks" -> "Max blocks"
            "min_turnovers" -> "Fewest turnovers"
            "max_fouls" -> "Max fouls"
            "max_sets_won" -> "Max sets won"
            "max_games_won" -> "Max games won"
            "max_aces" -> "Max aces"
            "min_double_faults" -> "Fewest double faults"
            "max_winners" -> "Max winners"
            "min_unforced_errors" -> "Fewest unforced errors"
            "max_break_points_won" -> "Max break points won"
            "max_net_points_won" -> "Max net points won"
            "max_digs" -> "Max digs"
            "min_official_time_sec" -> "Best official time"
            "max_runs_count" -> "Max runs"
            "max_speed_kmh" -> "Max speed"
            "max_routes_sent" -> "Max routes sent"
            "max_flashes" -> "Max flashes"
            "max_hits" -> "Max hits"
            "max_penalty_minutes" -> "Max penalty minutes"
            "max_tries" -> "Max tries"
            "max_meters_gained" -> "Max meters gained"
            "max_offloads" -> "Max offloads"
            "max_turnovers_won" -> "Max turnovers won"
            else -> snakeToTitle(m)
        }
    }

    fun formatValue(metric: String, value: Double?, label: String = ""): String {
        if (value == null) return "—"
        val m = metric.lowercase(Locale.US)
        if (m.endsWith("_kg") || m == "est_1rm_kg" || m == "max_weight_kg" || m == "best_set_volume_kg") {
            return String.format(Locale.US, "%.1f kg", value)
        }
        if (m.contains("reps") && !m.contains("sec")) {
            return "${value.roundToInt()} reps"
        }
        if (m == "max_hr" || m == "max_avg_hr") {
            return "${value.roundToInt()} bpm"
        }
        if (m == "longest_distance_km" || m == "max_total_distance_km") {
            return String.format(Locale.US, "%.1f km", value)
        }
        if (m == "max_speed_kmh") {
            return String.format(Locale.US, "%.1f km/h", value)
        }
        if (m == "max_elevation_m" || m == "max_vertical_drop_m" || m == "max_total_vertical_m") {
            return "${value.roundToInt()} m"
        }
        if (m == "max_cadence_rpm") {
            val unit = if (label.lowercase(Locale.US) == "rowerg") "spm" else "rpm"
            return "${value.roundToInt()} $unit"
        }
        if (m == "max_watts_avg") {
            return "${value.roundToInt()} W"
        }
        if (m == "fastest_pace_sec_per_km") {
            return paceString(value, "/km")
        }
        if (m == "min_split_sec_per_500m") {
            return paceString(value, "/500m")
        }
        if (m.endsWith("_sec") || m.contains("duration") || m.contains("time")) {
            return durationString(value)
        }
        if (m.endsWith("_km")) {
            return String.format(Locale.US, "%.1f km", value)
        }
        if (m.endsWith("_kmh")) {
            return String.format(Locale.US, "%.1f km/h", value)
        }
        if (m.endsWith("_m") && !m.contains("min")) {
            return "${value.roundToInt()} m"
        }
        if (m.startsWith("max_") || m.startsWith("min_")) {
            return "${value.roundToInt()}"
        }
        return String.format(Locale.US, "%.2f", value)
    }

    fun lowerIsBetter(metric: String): Boolean {
        val m = metric.lowercase(Locale.US)
        return m.startsWith("min_")
            || m.contains("pace")
            || m.contains("fastest")
            || m.contains("official_time")
            || m.contains("split")
            || m.contains("turnover")
            || m.contains("double_fault")
            || m.contains("unforced_error")
    }

    fun metricSortOrder(metric: String): Int {
        val order = listOf(
            "longest_distance_km", "max_total_distance_km",
            "longest_duration_sec", "max_moving_time_sec", "min_official_time_sec",
            "fastest_pace_sec_per_km", "min_split_sec_per_500m",
            "max_speed_kmh",
            "max_elevation_m", "max_vertical_drop_m", "max_total_vertical_m",
            "max_hr", "max_avg_hr",
            "max_weight_kg", "est_1rm_kg", "best_set_volume_kg", "max_reps",
            "max_cadence_rpm", "max_watts_avg",
            "max_goals", "max_assists", "max_points", "max_tries",
            "max_routes_sent", "max_flashes", "max_runs_count"
        )
        val m = metric.lowercase(Locale.US)
        val idx = order.indexOf(m)
        if (idx >= 0) return idx
        return 1000 + m.hashCode().mod(1000)
    }

    fun kindSortOrder(kind: String): Int = when (kind.lowercase(Locale.US)) {
        "strength" -> 0
        "cardio" -> 1
        "sport" -> 2
        else -> 3
    }

    private fun cardioTypeLabel(type: AddCardioActivity): String = when (type) {
        AddCardioActivity.RUN -> "Run"
        AddCardioActivity.WALK -> "Walk"
        AddCardioActivity.HIKE -> "Hike"
        AddCardioActivity.TREADMILL -> "Treadmill"
        AddCardioActivity.BIKE -> "Bike"
        AddCardioActivity.E_BIKE -> "E-Bike"
        AddCardioActivity.MTB -> "MTB"
        AddCardioActivity.INDOOR_CYCLING -> "Indoor cycling"
        AddCardioActivity.ROWERG -> "RowErg"
        AddCardioActivity.SWIM_POOL -> "Swim (pool)"
        AddCardioActivity.SWIM_OPEN_WATER -> "Swim (open water)"
    }

    private fun sportTypeLabel(type: AddSportType): String = when (type) {
        AddSportType.PADEL -> "Padel"
        AddSportType.TENNIS -> "Tennis"
        AddSportType.FOOTBALL -> "Football"
        AddSportType.BASKETBALL -> "Basketball"
        AddSportType.BADMINTON -> "Badminton"
        AddSportType.SQUASH -> "Squash"
        AddSportType.TABLE_TENNIS -> "Table Tennis"
        AddSportType.VOLLEYBALL -> "Volleyball"
        AddSportType.HANDBALL -> "Handball"
        AddSportType.HOCKEY -> "Hockey"
        AddSportType.RUGBY -> "Rugby"
        AddSportType.HYROX -> "Hyrox"
        AddSportType.SKI -> "Ski"
        AddSportType.CLIMBING -> "Climbing"
    }

    private fun sportIcon(sport: String): String = when (sport) {
        "padel", "tennis", "squash", "badminton", "table_tennis" -> "🎾"
        "football" -> "⚽️"
        "basketball" -> "🏀"
        "volleyball" -> "🏐"
        "rugby" -> "🏉"
        "hockey", "field_hockey" -> "🏑"
        "handball" -> "🤾‍♂️"
        "hyrox" -> "🔥"
        "ski" -> "⛷️"
        "climbing" -> "🧗"
        else -> "🏅"
    }

    private fun cardioIcon(activity: String): String = when (activity) {
        "run", "outdoor_run", "trail_run" -> "🏃‍♂️"
        "treadmill" -> "🏃‍♀️"
        "walk", "hike" -> "🚶‍♂️"
        "bike", "e_bike", "mtb", "cycling", "road_cycling" -> "🚴‍♂️"
        "indoor_cycling", "spinning" -> "🚲"
        "rowerg", "rowing" -> "🚣‍♂️"
        "swim_pool" -> "🏊‍♂️"
        "swim_open_water" -> "🌊"
        "cardio" -> "❤️‍🔥"
        else -> "❤️‍🔥"
    }

    private fun durationString(value: Double): String {
        val s = max(0, value.roundToInt())
        val h = s / 3600
        val mm = (s % 3600) / 60
        val ss = s % 60
        return if (h > 0) {
            String.format(Locale.US, "%d:%02d:%02d", h, mm, ss)
        } else {
            String.format(Locale.US, "%d:%02d", mm, ss)
        }
    }

    private fun paceString(value: Double, suffix: String): String {
        val s = max(1, value.roundToInt())
        return String.format(Locale.US, "%d:%02d %s", s / 60, s % 60, suffix)
    }

    private fun snakeToTitle(s: String): String =
        s.replace('_', ' ').replaceFirstChar { c ->
            if (c.isLowerCase()) c.titlecase(Locale.getDefault()) else c.toString()
        }
}

File: Liftr/PrFormatting.swift
import Foundation

enum PrFormatting {
    static func activityLabel(kind: String, label: String) -> String {
        let k = kind.lowercased()
        let wire = label.lowercased()
        if k == "cardio" {
            if let t = CardioActivityType(rawValue: wire) { return t.label }
            if wire == "cardio" { return "Cardio" }
            return snakeToTitle(wire)
        }
        if k == "sport" {
            if let t = SportType(rawValue: wire) { return t.label }
            return snakeToTitle(wire)
        }
        return label
    }

    static func activityIcon(kind: String, label: String) -> String {
        let k = kind.lowercased()
        let wire = label.lowercased()
        if k == "strength" { return "💪" }
        if k == "cardio" { return cardioIcon(for: wire) }
        if k == "sport" { return sportIcon(for: wire) }
        return ""
    }

    static func prettyMetricName(_ metric: String, kind: String, label: String) -> String {
        let m = metric.lowercased()
        switch m {
        case "max_hr", "max_avg_hr": return "Max HR"
        case "longest_duration_sec", "max_moving_time_sec": return "Longest duration"
        case "longest_distance_km", "max_total_distance_km": return "Longest distance"
        case "fastest_pace_sec_per_km": return "Fastest pace"
        case "max_elevation_m", "max_vertical_drop_m", "max_total_vertical_m": return "Max elevation"
        case "est_1rm_kg": return "Estimated 1RM"
        case "max_weight_kg": return "Max weight"
        case "best_set_volume_kg": return "Best set volume"
        case "max_reps": return "Max reps"
        case "max_cadence_rpm": return "Max cadence"
        case "max_watts_avg": return "Max watts"
        case "min_split_sec_per_500m": return "Best split"
        case "max_goals": return "Max goals"
        case "max_assists": return "Max assists"
        case "max_shots_on_target": return "Max shots on target"
        case "max_shots_on_goal": return "Max shots on goal"
        case "max_saves": return "Max saves"
        case "max_tackles", "max_tackles_made": return "Max tackles"
        case "max_yellow": return "Max yellow cards"
        case "max_red": return "Max red cards"
        case "max_points": return "Max points"
        case "max_rebounds": return "Max rebounds"
        case "max_steals": return "Max steals"
        case "max_blocks": return "Max blocks"
        case "min_turnovers": return "Fewest turnovers"
        case "max_fouls": return "Max fouls"
        case "max_sets_won": return "Max sets won"
        case "max_games_won": return "Max games won"
        case "max_aces": return "Max aces"
        case "min_double_faults": return "Fewest double faults"
        case "max_winners": return "Max winners"
        case "min_unforced_errors": return "Fewest unforced errors"
        case "max_break_points_won": return "Max break points won"
        case "max_net_points_won": return "Max net points won"
        case "max_digs": return "Max digs"
        case "min_official_time_sec": return "Best official time"
        case "max_runs_count": return "Max runs"
        case "max_speed_kmh": return "Max speed"
        case "max_routes_sent": return "Max routes sent"
        case "max_flashes": return "Max flashes"
        case "max_hits": return "Max hits"
        case "max_penalty_minutes": return "Max penalty minutes"
        case "max_tries": return "Max tries"
        case "max_meters_gained": return "Max meters gained"
        case "max_offloads": return "Max offloads"
        case "max_turnovers_won": return "Max turnovers won"
        default: return snakeToTitle(m)
        }
    }

    static func formatValue(metric: String, value: Double?, label: String = "") -> String {
        guard let v = value else { return "—" }
        let m = metric.lowercased()
        if m.hasSuffix("_kg") || m == "est_1rm_kg" || m == "max_weight_kg" || m == "best_set_volume_kg" {
            return String(format: "%.1f kg", v)
        }
        if m.contains("reps") && !m.contains("sec") {
            return "\(Int(v.rounded())) reps"
        }
        if m == "max_hr" || m == "max_avg_hr" {
            return "\(Int(v.rounded())) bpm"
        }
        if m == "longest_distance_km" || m == "max_total_distance_km" {
            return String(format: "%.1f km", v)
        }
        if m == "max_speed_kmh" {
            return String(format: "%.1f km/h", v)
        }
        if m == "max_elevation_m" || m == "max_vertical_drop_m" || m == "max_total_vertical_m" {
            return "\(Int(v.rounded())) m"
        }
        if m == "max_cadence_rpm" {
            let unit = label.lowercased() == "rowerg" ? "spm" : "rpm"
            return "\(Int(v.rounded())) \(unit)"
        }
        if m == "max_watts_avg" {
            return "\(Int(v.rounded())) W"
        }
        if m == "fastest_pace_sec_per_km" {
            return paceString(fromSeconds: v, suffix: "/km")
        }
        if m == "min_split_sec_per_500m" {
            return paceString(fromSeconds: v, suffix: "/500m")
        }
        if m.hasSuffix("_sec") || m.contains("duration") || m.contains("time") {
            return durationString(fromSeconds: v)
        }
        if m.hasSuffix("_km") {
            return String(format: "%.1f km", v)
        }
        if m.hasSuffix("_kmh") {
            return String(format: "%.1f km/h", v)
        }
        if m.hasSuffix("_m") && !m.contains("min") {
            return "\(Int(v.rounded())) m"
        }
        if m.hasPrefix("max_") || m.hasPrefix("min_") {
            return "\(Int(v.rounded()))"
        }
        return String(format: "%.2f", v)
    }

    static func lowerIsBetter(metric: String) -> Bool {
        let m = metric.lowercased()
        return m.hasPrefix("min_")
            || m.contains("pace")
            || m.contains("fastest")
            || m.contains("official_time")
            || m.contains("split")
            || m.contains("turnover")
            || m.contains("double_fault")
            || m.contains("unforced_error")
    }

    static func metricSortOrder(_ metric: String) -> Int {
        let order = [
            "longest_distance_km", "max_total_distance_km",
            "longest_duration_sec", "max_moving_time_sec", "min_official_time_sec",
            "fastest_pace_sec_per_km", "min_split_sec_per_500m",
            "max_speed_kmh",
            "max_elevation_m", "max_vertical_drop_m", "max_total_vertical_m",
            "max_hr", "max_avg_hr",
            "max_weight_kg", "est_1rm_kg", "best_set_volume_kg", "max_reps",
            "max_cadence_rpm", "max_watts_avg",
            "max_goals", "max_assists", "max_points", "max_tries",
            "max_routes_sent", "max_flashes", "max_runs_count"
        ]
        let m = metric.lowercased()
        if let idx = order.firstIndex(of: m) { return idx }
        return 1000 + m.hashValue % 1000
    }

    static func kindSortOrder(_ kind: String) -> Int {
        switch kind.lowercased() {
        case "strength": return 0
        case "cardio": return 1
        case "sport": return 2
        default: return 3
        }
    }

    private static func sportIcon(for sport: String) -> String {
        switch sport {
        case "padel", "tennis", "squash", "badminton", "table_tennis": return "🎾"
        case "football": return "⚽️"
        case "basketball": return "🏀"
        case "volleyball": return "🏐"
        case "rugby": return "🏉"
        case "hockey", "field_hockey": return "🏑"
        case "handball": return "🤾‍♂️"
        case "hyrox": return "🔥"
        case "ski": return "⛷️"
        case "climbing": return "🧗"
        default: return "🏅"
        }
    }

    private static func cardioIcon(for activity: String) -> String {
        switch activity {
        case "run", "outdoor_run", "trail_run": return "🏃‍♂️"
        case "treadmill": return "🏃‍♀️"
        case "walk", "hike": return "🚶‍♂️"
        case "bike", "e_bike", "mtb", "cycling", "road_cycling": return "🚴‍♂️"
        case "indoor_cycling", "spinning": return "🚲"
        case "rowerg", "rowing": return "🚣‍♂️"
        case "swim_pool": return "🏊‍♂️"
        case "swim_open_water": return "🌊"
        case "cardio": return "❤️‍🔥"
        default: return "❤️‍🔥"
        }
    }

    private static func durationString(fromSeconds secondsDouble: Double) -> String {
        let s = max(0, Int(secondsDouble.rounded()))
        let h = s / 3600
        let m = (s % 3600) / 60
        let sec = s % 60
        if h > 0 { return String(format: "%d:%02d:%02d", h, m, sec) }
        return String(format: "%d:%02d", m, sec)
    }

    private static func paceString(fromSeconds secondsDouble: Double, suffix: String) -> String {
        let s = max(1, Int(secondsDouble.rounded()))
        let m = s / 60
        let sec = s % 60
        return String(format: "%d:%02d %@", m, sec, suffix)
    }

    private static func snakeToTitle(_ s: String) -> String {
        s.replacingOccurrences(of: "_", with: " ").capitalized
    }
}
